### PR TITLE
487 add components to show application lifecycle data

### PR DESCRIPTION
**Pull request by neremic**
#487 
added the components and wrote tests

### CODE REVIEW

**Comment by kolja** [comment]
👍

File: client/lib/application/src/application-lifecycle/application-lifecycle.jsx [hunk 1]
```diff
@@ -1,25 +1,124 @@
 import React from 'react';
-/*eslint-disable */
+import PropsExposer from 'common/src/components/pure/PropsExposer.jsx';
+import ComboBox from 'common/src/components/pure/ComboBox.jsx';
+import ThreeColumns from 'common/src/components/pure/ThreeColumns.jsx';
```
Comment (prayerslayer) [inline client/lib/application/src/application-lifecycle/application-lifecycle.jsx]: Nitpick: For consistency I'd rather have these in `common/src/` without the `components` folder in between (as there are already React components there this distinction makes no sense).

File: client/lib/application/src/application-lifecycle/components/Error.jsx [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+
+const Error = (props) => {
```
Comment (prayerslayer) [inline client/lib/application/src/application-lifecycle/components/Error.jsx]: Maybe reuse the existing error component? https://github.com/zalando-stups/yourturn/blob/master/client/lib/common/src/error.jsx

File: client/lib/application/src/application-lifecycle/components/Loading.jsx [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import ThreeColumns from 'common/src/components/pure/ThreeColumns.jsx';
+import Icon from 'react-fa';
+
+const Loading = () => {
+    return (
+        <ThreeColumns
```
Comment (prayerslayer) [inline client/lib/application/src/application-lifecycle/components/Loading.jsx]: So if you want to reuse the Loading component, you always get three flex columns along with it?

File: client/lib/application/src/application-lifecycle/ApplicationLifeCycleComponent.jsx [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    }).isRequired,
+    params: React.PropTypes.shape({
+        applicationId: React.PropTypes.string
+    }).isRequired
+};
+
+function mapStateToProps(state) {
+    return {
+        aliceStore: state.alice,
+        kioStore: bindGettersToState(state.kio, KioGetter)
```
Comment (prayerslayer) [inline client/lib/application/src/application-lifecycle/ApplicationLifeCycleComponent.jsx]: I'd like to not have this pattern in new code, as it was just introduced for quicker refactoring when switching from Flummox to Redux. Please pass relevant data directly as props. (Same goes for actions, no need for `bindActionCreators` — pass the actions as props.)

File: client/lib/application/src/application-lifecycle/components/Loading.jsx [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+            leftChildren   = {<div></div>}
+            middleChildren = {<div><Icon pulse size='5x' name='spinner' /> Loading</div>}
```
Comment (prayerslayer) [inline client/lib/application/src/application-lifecycle/components/Loading.jsx]: The spinner icon in YT is `circle-o-notch` :)

File: client/lib/common/asset/less/brush.less [hunk 6]
```diff
@@ -0,0 +1,10 @@
+.brush {
+  .extent {
+    fill-opacity: .125;
+    shape-rendering: crispEdges;
+    stroke: #000000;
```
Comment (prayerslayer) [inline client/lib/common/asset/less/brush.less]: Please reuse existing colors if possible: https://github.com/zalando-stups/yourturn/blob/master/client/lib/common/asset/less/variables.less#L14

File: client/lib/common/src/components/pure/Brush.jsx [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import React from 'react'
+
+import { Brush } from 'react-d3-components';
```
Comment (prayerslayer) [inline client/lib/common/src/components/pure/Brush.jsx]: import { Brush as D3Brush } from 'react-d3-components'

Then you don't need the `ZBrush` :wink:

File: client/lib/common/src/components/pure/PropsExposer.jsx [hunk 8]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+
+var PropsExposer = (Wrapped, callback) => class PropsExposer extends React.Component {
```
Comment (prayerslayer) [inline client/lib/common/src/components/pure/PropsExposer.jsx]: Is this one still used? I thought we had an alternative.

File: client/lib/common/src/data/services.js [hunk 9]
```diff
@@ -35,6 +36,10 @@ const SERVICES = {
         url: PIERONE_BASE_URL,
         root: '/v1/search',
         searchQuery: 'q'
+    },
```
Comment (prayerslayer) [inline client/lib/common/src/data/services.js]: Ya, we don't need this here... Was a lame attempt to abstract link construction to the other services.

File: client/webpack.config.js [hunk 10]
```diff
@@ -60,7 +61,7 @@ module.exports = {
     module: {
         loaders: [
             {test: /\.jsx?$/, exclude: /node_modules/, loaders: ['react-hot', 'babel']},
-            {test: /\.less$/, exclude: /node_modules/, loaders: ['style', 'css', 'autoprefixer', 'less']},
+            {test: /\.less$/, loaders: ['style', 'css', 'autoprefixer', 'less']},
```
Comment (prayerslayer) [inline client/webpack.config.js]: Would the `null` loader also be sufficient?

File: package.json [hunk 11]
```diff
@@ -12,7 +12,9 @@
     "lodash": "cd client && gulp lodash",
     "test-server": "cd server && npm install && mocha -r test/mocha-globals.js test/**/*.test.js",
     "test": "cd client && webpack --config webpack.test.config.js && mocha -r mocha-globals.js dist/test.js",
-    "test-coverage": "cd client && webpack --config webpack.test.config.js && istanbul cover ../node_modules/mocha/bin/_mocha --report lcovonly -- -r mocha-globals.js dist/test.js && cat ./coverage/lcov.info | ../node_modules/coveralls/bin/coveralls.js"
+    "test2": "mocha --reporter spec tools/testSetup.js \"client/lib/application/test/**/*.test.jsx\"",
```
Comment (prayerslayer) [inline package.json]: Maybe you find a better name than `test2`? :grin:

File: package.json [hunk 12]
```diff
@@ -36,12 +38,16 @@
...
     "lz-string": "1.4.4",
+    "moment": "2.11.1",
     "node-uuid": "1.4.7",
     "object-assign": "4.0.1",
     "raf": "3.1.0",
+    "rd3": "^0.7.0",
     "react": "0.14.7",
     "react-copy-to-clipboard": "3.0.4",
-    "react-date-picker": "4.0.10",
+    "react-d3-components": "^0.6.4",
```
Comment (prayerslayer) [inline package.json]: Can you pin the dependencies to a specific version just like the others? Thanks.

File: package.json (hunk 11 above)
Comment (neremic) [inline package.json]: Forgot to remove this line. Even better than a better name

File: client/lib/application/src/application-lifecycle/components/Loading.jsx (hunk 3 above)
Comment (neremic) [inline client/lib/application/src/application-lifecycle/components/Loading.jsx]: All components under application-lifecycle are not for reuse (those reside in common) but merely only to split up the application lifecycle main component into smaller parts

File: client/lib/application/src/application-lifecycle/application-lifecycle.jsx (hunk 1 above)
Comment (neremic) [inline client/lib/application/src/application-lifecycle/application-lifecycle.jsx]: I'd really like to introduce the components folder in common/src/ as there is already a data folder. in long term I figured to move all components that are on top level to a components sub folder.

File: client/lib/application/src/application-lifecycle/components/Error.jsx (hunk 2 above)
Comment (neremic) [inline client/lib/application/src/application-lifecycle/components/Error.jsx]: ok. will try to

File: client/lib/application/src/application-lifecycle/components/Loading.jsx (hunk 5 above)
Comment (neremic) [inline client/lib/application/src/application-lifecycle/components/Loading.jsx]: will change to circle-o-notch.
as for pulse in conjunction with spinner see:
http://fontawesome.io/examples/ in "animated icons" last icon :)

File: client/lib/common/asset/less/brush.less (hunk 6 above)
Comment (neremic) [inline client/lib/common/asset/less/brush.less]: yep!

File: client/lib/common/src/components/pure/Brush.jsx (hunk 7 above)
Comment (neremic) [inline client/lib/common/src/components/pure/Brush.jsx]: I feel stupdig now. will change this o/c

File: client/lib/common/src/components/pure/PropsExposer.jsx (hunk 8 above)
Comment (neremic) [inline client/lib/common/src/components/pure/PropsExposer.jsx]: alternative didnt work for me. so I reverted to proven concept :)

File: client/lib/application/src/application-lifecycle/components/Loading.jsx (hunk 3 above)
Comment (prayerslayer) [inline client/lib/application/src/application-lifecycle/components/Loading.jsx]: Ok, but I think `Loading` is not that special, so we can possibly split it up? Ie. one for reuse in `common` and one with columns in `application-lifecycle`?

File: client/lib/application/src/application-lifecycle/application-lifecycle.jsx (hunk 1 above)
Comment (prayerslayer) [inline client/lib/application/src/application-lifecycle/application-lifecycle.jsx]: OK, then refactor and move the existing ones into `components`?

**Comment by kolja** [comment]
👍

File: client/lib/application/src/application-lifecycle/ApplicationLifeCycleComponent.jsx [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+        fetchInstanceCount: React.PropTypes.func
+    }).isRequired,
+    params: React.PropTypes.shape({
+        applicationId: React.PropTypes.string
+    }).isRequired
+};
+
+function mapStateToProps(state) {
+    return {
+        aliceStore: state.alice,
```
Comment (prayerslayer) [inline client/lib/application/src/application-lifecycle/ApplicationLifeCycleComponent.jsx]: We could split this up as well (looks weird) but never mind.

File: client/lib/application/src/application-lifecycle/components/Toolbar.jsx [hunk 14]
```diff
@@ -0,0 +1,59 @@
+import React from 'react';
+import moment from 'moment';
+
+import Brush from 'common/src/components/pure/Brush.jsx';
+import DateSelector, {STYLE_RIGHT} from 'common/src/components/functional/DateSelector.jsx';
+import ThreeColumns from 'common/src/components/pure/ThreeColumns.jsx';
+
+const DATE_FORMAT = 'Do [of] MMM YY';
```
Comment (prayerslayer) [inline client/lib/application/src/application-lifecycle/components/Toolbar.jsx]: Uh didn't catch that one last time. We format dates with this: https://github.com/zalando-stups/yourturn/blob/master/client/lib/common/src/config.js#L4

File: client/lib/common/src/components/functional/AutoWidthContainer.jsx [hunk 15]
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
+import Dimensions from 'react-dimensions';
```
Comment (prayerslayer) [inline client/lib/common/src/components/functional/AutoWidthContainer.jsx]: Now we have two libraries doing the same thing? :)

File: client/lib/common/src/components/pure/Spinner.jsx [hunk 16]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+
+import Icon from 'react-fa';
+
+const Spinner = (props) => {
+    return (
+        <Icon spin size={props.size} name='spinner' />
```
Comment (prayerslayer) [inline client/lib/common/src/components/pure/Spinner.jsx]: `circle-o-notch`? :)

**Comment by prayerslayer** [comment]
Please change the spinner icon, I can live with the rest for now.

**Comment by neremic** [comment]
:+1:

**Comment by prayerslayer** [comment]
👍

**Comment by neremic** [comment]
:+1:

**Pull request merged**